kubernetes/resource_kubernetes_role.go

```diff
@@ -0,0 +1,227 @@
+							Elem:        &schema.Schema{Type: schema.TypeString},
+							Set:         schema.HashString,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceKubernetesRoleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	rules := expandRules(d.Get("rule").([]interface{}))
+
+	role := v1.Role{
+		ObjectMeta: metadata,
+		Rules:      rules,
+	}
+	log.Printf("[INFO] Creating new role: %#v", role)
+	out, err := conn.RbacV1().Roles(metadata.Namespace).Create(&role)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Submitted new role: %#v", out)
+	d.SetId(buildId(out.ObjectMeta))
+
+	return resourceKubernetesRoleRead(d, meta)
+}
+
+func resourceKubernetesRoleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Reading role %s", name)
+	role, err := conn.RbacV1().Roles(namespace).Get(name, meta_v1.GetOptions{})
+	if err != nil {
+		log.Printf("[DEBUG] Received error: %#v", err)
+		return err
+	}
+
+	log.Printf("[INFO] Received role: %#v", role)
+	err = d.Set("metadata", flattenMetadata(role.ObjectMeta))
+	if err != nil {
+		return err
+	}
+
+	err = d.Set("rule", flattenRules(role.Rules))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceKubernetesRoleUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, _, err := idParts(d.Id())
+	if err != nil {
+		return err
+	}
+
+	metadata := expandMetadata(d.Get("metadata").([]interface{}))
+	rules := expandRules(d.Get("rule").([]interface{}))
+
+	role := v1.Role{
+		ObjectMeta: metadata,
+		Rules:      rules,
+	}
+
+	out, err := conn.RbacV1().Roles(namespace).Update(&role)
+	if err != nil {
+		return fmt.Errorf("Failed to update role: %s", err)
+	}
+	log.Printf("[INFO] Submitted updated role: %#v", out)
+	d.SetId(buildId(out.ObjectMeta))
+
+	return resourceKubernetesRoleRead(d, meta)
+}
+
+func resourceKubernetesRoleDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Deleting role: %#v", name)
+	err = conn.RbacV1().Roles(namespace).Delete(name, &meta_v1.DeleteOptions{})
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[INFO] Role %s deleted", name)
+
+	d.SetId("")
+	return nil
+}
+
+func resourceKubernetesRoleExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	conn := meta.(*kubernetes.Clientset)
+
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return false, err
+	}
+
+	log.Printf("[INFO] Checking role %s", name)
+	_, err = conn.RbacV1().Roles(namespace).Get(name, meta_v1.GetOptions{})
+	if err != nil {
+		if statusErr, ok := err.(*errors.StatusError); ok && statusErr.ErrStatus.Code == 404 {
```

Comment:
The client-go SDK has helpers for determining the kind of error that was received.
Can you please refactor this line to make use of those instead of checking directly against `404`?

See the example from `StatefulSets` which I mentioned before.
The exhaustive list is here: https://godoc.org/k8s.io/apimachinery/pkg/api/errors